File: examples/crypto/secp256k1/signature-validation.ts
/**
 * Signature validation and edge cases
 *
 * Demonstrates:
 * - Signature component validation
 * - Low-s malleability protection
 * - Invalid signature detection
 * - Edge case handling
 * - Security best practices
 */

import * as Secp256k1 from "../../../src/crypto/Secp256k1/index.js";
import { keccak256 } from "../../../src/primitives/Hash/BrandedHash/keccak256.js";

// Generate test keypair
const privateKeyBytes = new Uint8Array(32);
crypto.getRandomValues(privateKeyBytes);
const privateKey = `0x${Buffer.from(privateKeyBytes).toString("hex")}`;
const publicKey = Secp256k1.derivePublicKey(privateKey as any);

// Create valid signature
const message = "Test message";
const messageBytes = new TextEncoder().encode(message);
const messageHash = keccak256(messageBytes);
const validSignature = Secp256k1.sign(messageHash, privateKey);

// Validate signature structure
const isValid = Secp256k1.isValidSignature(validSignature);

// Verify signature
const verifies = Secp256k1.verify(validSignature, messageHash, publicKey);
const invalidR = {
	r: new Uint8Array(32) as any, // All zeros
	s: validSignature.s,
	v: validSignature.v,
};

const rIsValid = Secp256k1.isValidSignature(invalidR);

try {
	const rVerifies = Secp256k1.verify(invalidR, messageHash, publicKey);
} catch (error) {}
const invalidS = {
	r: validSignature.r,
	s: new Uint8Array(32).fill(0xff) as any, // All 0xFF > curve order
	v: validSignature.v,
};

const sIsValid = Secp256k1.isValidSignature(invalidS);

try {
	const sVerifies = Secp256k1.verify(invalidS, messageHash, publicKey);
} catch (error) {}
const wrongLengthR = {
	r: new Uint8Array(16) as any, // Too short
	s: validSignature.s,
	v: validSignature.v,
};

try {
	const lengthVerifies = Secp256k1.verify(wrongLengthR, messageHash, publicKey);
} catch (error) {}

const CURVE_ORDER = BigInt(
	"0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141",
);

// Convert s to BigInt
const sValue = BigInt(`0x${Buffer.from(validSignature.s).toString("hex")}`);
const halfOrder = CURVE_ORDER / 2n;

// Valid key
const validKey = new Uint8Array(32);
validKey[31] = 42;

// Zero key (invalid)
const zeroKey = new Uint8Array(32);

// Key >= n (invalid)
const tooLargeKey = new Uint8Array([
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xff,
	0xfe,
	0xba,
	0xae,
	0xdc,
	0xe6,
	0xaf,
	0x48,
	0xa0,
	0x3b,
	0xbf,
	0xd2,
	0x5e,
	0x8c,
	0xd0,
	0x36,
	0x41,
	0x41, // n (curve order) - invalid
]);

// Random bytes (likely not on curve)
const randomBytes = new Uint8Array(64);
crypto.getRandomValues(randomBytes);

// Wrong length
const wrongLength = new Uint8Array(32);

File: examples/crypto/secp256k1/signature-validation.zig
/// Signature validation and edge cases
///
/// Demonstrates:
/// - Signature component validation
/// - Low-s malleability protection
/// - Invalid signature detection
/// - Edge case handling
/// - Security best practices

const std = @import("std");
const crypto = @import("crypto");
const primitives = @import("primitives");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("=== Signature Validation ===\n\n", .{});

    // Generate test keypair
    var private_key: [32]u8 = undefined;
    crypto.secp256k1.generatePrivateKey(&private_key);
    const public_key = try crypto.secp256k1.derivePublicKey(private_key);

    // Create valid signature
    const message = "Test message";
    const message_hash = try primitives.Hash.keccak256String(allocator, message);
    defer allocator.free(message_hash);

    const valid_signature = try crypto.secp256k1.sign(message_hash[0..32].*, private_key);

    std.debug.print("1. Valid Signature\n", .{});
    std.debug.print("   Message: {s}\n", .{message});
    std.debug.print("   r: {s}...\n", .{std.fmt.fmtSliceHexLower(valid_signature.r[0..8])});
    std.debug.print("   s: {s}...\n", .{std.fmt.fmtSliceHexLower(valid_signature.s[0..8])});
    std.debug.print("   v: {d}\n", .{valid_signature.v});

    // Validate signature structure
    const is_valid = crypto.secp256k1.isValidSignature(valid_signature);
    std.debug.print("   Valid structure: {s}\n", .{if (is_valid) "✓ Yes" else "✗ No"});

    // Verify signature
    const verifies = try crypto.secp256k1.verify(valid_signature, message_hash[0..32].*, public_key);
    std.debug.print("   Verifies: {s}\n", .{if (verifies) "✓ Yes" else "✗ No"});

    // 2. Invalid r component (all zeros)
    std.debug.print("\n2. Invalid r Component (r = 0)\n", .{});
    var invalid_r = valid_signature;
    invalid_r.r = [_]u8{0} ** 32;

    const r_is_valid = crypto.secp256k1.isValidSignature(invalid_r);
    std.debug.print("   Valid structure: {s}\n", .{if (!r_is_valid) "✗ No (expected)" else "✓ Yes"});

    // 3. Invalid s component (too large)
    std.debug.print("\n3. Invalid s Component (s >= n)\n", .{});
    var invalid_s = valid_signature;
    invalid_s.s = [_]u8{0xff} ** 32;

    const s_is_valid = crypto.secp256k1.isValidSignature(invalid_s);
    std.debug.print("   Valid structure: {s}\n", .{if (!s_is_valid) "✗ No (expected)" else "✓ Yes"});

    // 4. Low-s malleability check
    std.debug.print("\n4. Low-s Malleability Protection\n", .{});
    std.debug.print("   Signature s value: {s}\n", .{std.fmt.fmtSliceHexLower(&valid_signature.s)});
    std.debug.print("   Auto-normalized: ✓ Yes (RFC 6979 + low-s enforcement)\n", .{});

    // 5. Private key validation
    std.debug.print("\n5. Private Key Validation\n", .{});

    // Valid key
    var valid_key: [32]u8 = [_]u8{0} ** 32;
    valid_key[31] = 42;
    std.debug.print("   Valid key (42): {s}\n", .{if (crypto.secp256k1.isValidPrivateKey(valid_key)) "✓ Yes" else "✗ No"});

    // Zero key (invalid)
    const zero_key: [32]u8 = [_]u8{0} ** 32;
    std.debug.print("   Zero key: {s}\n", .{if (!crypto.secp256k1.isValidPrivateKey(zero_key)) "✗ No (expected)" else "✓ Yes"});

    // Key >= n (invalid)
    const too_large_key: [32]u8 = .{
        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe,
        0xba, 0xae, 0xdc, 0xe6, 0xaf, 0x48, 0xa0, 0x3b,
        0xbf, 0xd2, 0x5e, 0x8c, 0xd0, 0x36, 0x41, 0x41,
    };
    std.debug.print("   Key >= n: {s}\n", .{if (!crypto.secp256k1.isValidPrivateKey(too_large_key)) "✗ No (expected)" else "✓ Yes"});

    // 6. Public key validation
    std.debug.print("\n6. Public Key Validation\n", .{});

    // Valid key (derived)
    std.debug.print("   Valid derived key: {s}\n", .{if (crypto.secp256k1.isValidPublicKey(public_key)) "✓ Yes" else "✗ No"});

    // Random bytes (likely not on curve)
    var random_bytes: [64]u8 = undefined;
    std.crypto.random.bytes(&random_bytes);
    std.debug.print("   Random bytes: {s}\n", .{if (!crypto.secp256k1.isValidPublicKey(random_bytes)) "✗ No (expected)" else "✓ Yes"});

    // 7. Security recommendations
    std.debug.print("\n7. Security Recommendations\n", .{});
    std.debug.print("   ✓ Always validate signatures before use\n", .{});
    std.debug.print("   ✓ Check signature components (r, s) are in valid range\n", .{});
    std.debug.print("   ✓ Verify low-s to prevent malleability\n", .{});
    std.debug.print("   ✓ Validate public keys are on the curve\n", .{});
    std.debug.print("   ✓ Never reuse nonces (use RFC 6979)\n", .{});
    std.debug.print("   ✓ Use constant-time implementations\n", .{});
    std.debug.print("   ✓ Protect private keys with hardware security\n", .{});
}
